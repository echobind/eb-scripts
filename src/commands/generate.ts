import { Command, flags } from "@oclif/command";
import { execSync } from "child_process";
import {
  getTemplateLocation,
  pathWhereScriptIsRunning,
  rootDirectory
} from "../utils/";

export const DEFAULT_COMPONENT_NAME = "MyNewComponent";
const DEFAULT_TEMPLATE_NAME = "react-component";
const DEFAULT_PATH = "src/components";

/**
 * @description valid template type
 * @example react-component
 * @todo eventually, we'll add other template types
 */
type TemplateType = "react-component";

const validTemplateTypes: TemplateType[] = ["react-component"];

export default class Generate extends Command {
  static description = "generates new files";

  static examples = [
    `$ eb-scripts generate react-component -n MyNewComponent -p src/components
    Loaded templates: _templates
    added: src/MyNewComponent.js
`
  ];

  static flags = {
    help: flags.help({ char: "h" }),
    // flag with a value (-t, --template=VALUE)
    // name should correspond with one of the following names in /_templates
    template: flags.string({
      char: "t",
      description: "template to use",
      options: validTemplateTypes
    }),
    // flag with a value (-n, --name=VALUE)
    name: flags.string({
      char: "n",
      description: "name to print",
      default: DEFAULT_COMPONENT_NAME
    }),
    // flag with a value (-p, --path=VALUE)
    path: flags.string({
      char: "p",
      default: DEFAULT_PATH,
      description: "path to where you want the files to go"
    })
  };

  static args = [
    {
      name: "templateName",
      description: "the template you want to use",
      required: true,
      options: ["react-component"] as TemplateType[]
    }
  ];

  async run() {
    const { flags, args } = this.parse(Generate);
    // Assert flags.template as a TemplateType, so TS knows it should be a TemplateType
    const template =
      (args.templateName as TemplateType) || DEFAULT_TEMPLATE_NAME;
    const name = flags.name;
    // The path where the files will go when generated
    const path = flags.path;
    const templateLocation = getTemplateLocation();
    const templatePath = `HYGEN_TMPLS=${templateLocation}`;

    this.log(
      `Generating new component using template ${template} at ${pathWhereScriptIsRunning}/${path}`
    );

    // Generate template
    execSync(
      `${templatePath} yarn gen ${template} new ${name} --root=${pathWhereScriptIsRunning} --path=${path}`,
      { cwd: rootDirectory, stdio: "inherit" }
    );
  }
}
