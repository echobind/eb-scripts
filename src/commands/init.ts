import { Command, flags } from "@oclif/command";
import { execSync } from "child_process";
import * as fs from "fs";

const DEFAULT_PROJECT_NAME = "react";

/**
 * @description a script that is used in the `package.json` scripts
 * @example "g:component": "eb-scripts generate -t react-component -n"
 */
type Scripts = {
  /* A script */
  [script in string]: string;
};

/**
 * @description valid project type
 * @example `react`
 * @todo eventually, we'll add others like `react-native`, `react-typescript`, `node`, etc.
 */
type ProjectType = "react";

/**
 * @description the available options for generate scripts that may be added to a user's `package.json`
 */
type ScriptByProject = {
  [project in ProjectType]: Scripts;
};

// Grab the path of the user's project
const pathWhereScriptIsRunning = process.cwd();
const scriptsByProject: ScriptByProject = {
  react: {
    "g:component": "eb-scripts generate -t react-component -n"
  }
};

// Use the keys from scriptsByProject as valid projec types
const validProjectTypes: ProjectType[] = Object.keys(
  scriptsByProject
) as ProjectType[];
export default class Init extends Command {
  static description =
    "initializes project by installing `eb-scripts` and adding scripts to `package.json`";

  static examples = [`$ npx eb-scripts init -p react`];

  static flags = {
    help: flags.help({ char: "h" }),
    // flag with a value (--project=VALUE)
    project: flags.string({
      description: "language/framework of project",
      options: validProjectTypes
    })
  };

  static args = [
    {
      name: "project",
      required: true,
      description: "The language or framework of the project",
      default: "react",
      options: validProjectTypes
    }
  ];

  async run() {
    const { flags, args } = this.parse(Init);
    const project = args.project || flags.project || DEFAULT_PROJECT_NAME;
    // Grab their package Json
    const packageJsonLocation = `${pathWhereScriptIsRunning}/package.json`;
    const packageJson = require(packageJsonLocation);
    const theirScripts = packageJson.scripts || {};
    const validProjectTypes = Object.keys(scriptsByProject);
    const isValidProject = validProjectTypes.includes(project);

    // Check if project is valid
    if (!isValidProject) {
      console.error(`
        Error: invalid project type.
        Please use one of the following:
        ${validProjectTypes}
        `);
      return;
    }

    this.log(`Initializing new ${project} project`);

    // Otherwise, it is valid, so we tell TypeScript to assert it as a ProjectType
    const projectScripts: Scripts = scriptsByProject[project as ProjectType];

    // Combine their scripts with our scripts
    const updatedScripts = { ...projectScripts, ...theirScripts };
    packageJson.scripts = updatedScripts;

    this.log(`Writing to package.json scripts...`);
    fs.writeFileSync(packageJsonLocation, JSON.stringify(packageJson, null, 2));

    // *******************
    // Install `eb-scripts`
    // We do this so that the scripts we added to their package.json
    // work correctly.
    // *******************

    this.log(`Installing eb-scripts as a devDependency`);

    const yarnLockPath = pathWhereScriptIsRunning + "/yarn.lock";
    const hasYarnLock = fs.existsSync(yarnLockPath);

    // Check if they're using yarn
    if (hasYarnLock) {
      execSync("yarn add --dev eb-scripts", { cwd: pathWhereScriptIsRunning });
    } else {
      execSync("npm add --save-dev eb-scripts", {
        cwd: pathWhereScriptIsRunning
      });
    }

    this.log(`Done initializing.`);
  }
}
