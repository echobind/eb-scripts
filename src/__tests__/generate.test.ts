import * as path from "path";
import * as fse from "fs-extra";
import { execSync } from "child_process";
import { DEFAULT_COMPONENT_NAME } from "../commands/generate";

let root = process.cwd();
let tempRoot = path.join(`${root}/src`, "/components");

describe("The `generate` command", () => {
  beforeAll(async () => {
    // create our temporary directory
    try {
      await fse.ensureDir(tempRoot);
    } catch (error) {
      console.error(error);
    }
  });

  afterAll(async () => {
    // delete our temporary directory
    try {
      await fse.remove(tempRoot);
    } catch (err) {
      console.error(err);
    }
  });

  it("expects to find a temporary directory in src/components", async () => {
    const tempDirectoryExists = await fse.pathExists(tempRoot);

    expect(tempDirectoryExists).toBe(true);
  });

  it("throws an error if you don't pass any args", async () => {
    const generateCommand = () =>
      execSync(`./bin/run generate`, {
        cwd: root
      });

    expect(generateCommand).toThrowErrorMatchingSnapshot();
  });

  it("works with a arg of a valid template and uses the default src/components path flag", async () => {
    const componentName = "TestComponent";
    const componentFolderPath = `${tempRoot}/${componentName}`;

    execSync(`./bin/run generate react-component -n ${componentName}`, {
      cwd: root
    });

    await expect(newComponentFolderExists).toBe(true);
    // expect(componentIndexExists).toBe(true);
    done();
  });

  it("throws an error if you don't pass any args", async () => {
    const generateCommand = () =>
      execSync(`./bin/run generate`, {
        cwd: root
      });

    expect(generateCommand).toThrowErrorMatchingSnapshot();
  });

  it("works with a arg of a valid template and uses the default src/components path flag", async () => {
    const componentName = "TestComponent";
    const componentFolderPath = `${tempRoot}/${componentName}`;

    execSync(`./bin/run generate react-component -n ${componentName}`, {
      cwd: root
    });

    const newComponentFolderExists = await fse.pathExists(componentFolderPath);
    const componentIndexExists = fse.existsSync(
      `${componentFolderPath}/index.js`
    );

    expect(newComponentFolderExists).toBe(true);
    expect(componentIndexExists).toBe(true);
  });

  it("works with the react-typescript template and uses the default src/components path flag", async () => {
    const componentName = "TypeScriptComponent";
    const componentFolderPath = `${tempRoot}/${componentName}`;

    execSync(
      `./bin/run generate react-typescript-component -n ${componentName}`,
      {
        cwd: root
      }
    );

    const newComponentFolderExists = await fse.pathExists(componentFolderPath);
    const componentIndexExists = fse.existsSync(
      `${componentFolderPath}/index.ts`
    );

    const componentTsxExists = fse.existsSync(
      `${componentFolderPath}/${componentName}.tsx`
    );
    expect(newComponentFolderExists).toBe(true);
    expect(componentIndexExists).toBe(true);
    expect(componentTsxExists).toBe(true);
  });

  it("uses the default templates if none in the users directory and the default src/components path flag", async () => {
    const componentName = "DefaultTemplateComponent";
    const componentFolderPath = `${tempRoot}/${componentName}`;

    execSync(`./bin/run generate react-component -n ${componentName}`, {
      cwd: root
    });

    const newComponentFolderExists = await fse.pathExists(componentFolderPath);
    const componentIndexExists = fse.existsSync(
      `${componentFolderPath}/index.js`
    );

    expect(newComponentFolderExists).toBe(true);
    expect(componentIndexExists).toBe(true);
  });

  it("throws an error when you pass an invalid flag", () => {
    const generateCommand = () =>
      execSync(`./bin/run generate fake-component -n FakeComponent`, {
        cwd: root
      });

  it("throws an error when you pass an invalid flag", () => {
    const generateCommand = () =>
      execSync(`./bin/run generate fake-component -n FakeComponent`, {
        cwd: root
      });

    expect(generateCommand).toThrowErrorMatchingSnapshot();
  });

  it("uses the path flag if passed", async () => {
    const componentName = "PathFlagComponent";
    const customPath = "src/components/other";
    const componentFolderPath = `${customPath}/${componentName}`;

    execSync(
      `./bin/run generate react-component -n ${componentName} -p ${customPath}`,
      {
        cwd: root
      }
    );

    const newComponentFolderExists = await fse.pathExists(componentFolderPath);
    const componentIndexExists = fse.existsSync(
      `${componentFolderPath}/index.js`
    );

    expect(newComponentFolderExists).toBe(true);
    expect(componentIndexExists).toBe(true);
  });

  it("uses the users template if they have one", async () => {
    const pathToNewTemplate = `${root}/_templates/react-component/new`;
    const newComponentName = "Test";
    // Make a react-component template
    await fse.ensureDir(pathToNewTemplate);
    // Add template to _templates/react-component/new
    await fse.copyFileSync(
      `${root}/src/__tests__/helpers/componentTestTemplate.ejs.t`,
      `${pathToNewTemplate}/component.ejs.t`
    );

    const newTemplateFolderExists = await fse.pathExists(pathToNewTemplate);
    const newTemplateFileExists = fse.existsSync(
      `${pathToNewTemplate}/component.ejs.t`
    );

    // Check that it made our template and new file
    expect(newTemplateFolderExists).toBe(true);
    expect(newTemplateFileExists).toBe(true);

    execSync(`./bin/run generate react-component -n ${newComponentName}`, {
      cwd: root
    });

    // Using the template we created, we expect to see a new component
    const newGeneratedComponentExists = fse.existsSync(
      `${tempRoot}/${newComponentName}.jsx`
    );

    // And no folder for the component
    const newGeneratedComponentFolderExists = await fse.pathExists(
      `${tempRoot}/${newComponentName}`
    );

    expect(newGeneratedComponentExists).toBe(true);
    expect(newGeneratedComponentFolderExists).toBe(false);
    // Remove the /_templates so it doesn't cause side effects
    await fse.remove(`${root}/_templates`);
  });
});

    expect(newComponentFolderExists).toBe(true);
    expect(componentIndexExists).toBe(true);
  });
});
