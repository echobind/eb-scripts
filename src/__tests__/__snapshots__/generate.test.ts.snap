// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The \`generate\` command throws an error if you don't pass any args 1`] = `
"Command failed: ./bin/run generate
 ›   Error: Missing 1 required arg:
 ›   templateName  the template you want to use
 ›   See more help with --help
"
`;

exports[`The \`generate\` command throws an error when you pass an invalid flag 1`] = `
"Command failed: ./bin/run generate fake-component -n FakeComponent
 ›   Error: Expected fake-component to be one of: react-component,
 ›   react-typescript-component, react-native-typescript-component,
 ›   react-native-typescript-screen, react-native-e2e, util-typescript
 ›   See more help with --help
"
`;
